[Quant][fx] Create new FX-based LSTM reference module

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96343

Summary: The previous LSTM reference module implementation did
not handle dtypes other than quint8 correctly. This is because
the internal LSTM custom module quantization used eager mode,
which did not insert the q-dq ops properly. E.g., we want the
following reference quantized model:

```
[dq -> linear1_fp32 -> q_to_qint32] -> dq -> q_to_quint8 ->
  [dq - linear2_fp32 -> q_to_quint8] -> dq -> ...
```

This requires two sets of `q - dq` pairs between two adjacent
ops that have different dtypes (linear1 and linear2). However,
these `q - dq` pairs were not inserted in the old flow, because
eager mode required users to insert Quant/DeQuantStubs manually.

This commit changes the internal LSTM custom module quantization
to use FX graph mode quantization, which automatically inserts
the `q - dq` ops that convert the dtypes between adjacent ops
correctly. However, using FX graph mode quantization here comes
with its own set of challenges that required some hacks to get
the end-to-end flow to work. These hacks are detailed in the
comments in the util functions.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_static_lstm_with_custom_fixed_qparams

This commit also updates the corresponding test to verify the
dtypes as well as the qparams in the reference quantized graph.
This test case should serve as an example for users to set up
their own LSTM reference module flows.

Reviewers: vkuzo, supriyar, jcaip

Subscribers: vkuzo, supriyar, jcaip

andrewor14 committed:
[Quant][fx] Create new FX-based LSTM reference module

Summary: The previous reference LSTM module implementation did
not handle dtypes other than quint8 correctly. This was because
the internal LSTM custom module quantization used eager mode,
which did not insert the q-dq ops properly. E.g., we want the
following reference quantized graph:

```
[dq -> linear1_fp32 -> q_to_qint32] -> dq -> q_to_quint8 ->
  [dq - linear2_fp32 -> q_to_quint8] -> dq -> ...
```

This requires two sets of `q - dq` pairs between two adjacent
ops that have different dtypes (linear1 and linear2). These
`q - dq` pairs were not inserted in the old flow, because eager
mode did not do this automatically.

This commit changes the internal LSTM custom module quantization
to use FX graph mode quantization, which does automatically
insert the `q - dq` ops that convert the dtypes between adjacent
ops correctly. However, using FX graph mode quantization here
comes with its own set of challenges that required some hacks to
get the end-to-end flow to work. These challenges are detailed
in the test comments.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_static_lstm_with_custom_fixed_qparams

This commit also updates the corresponding test to verify the
dtypes as well as the qparams in the reference quantized graph.
This test case should serve as an example for users to set up
their own LSTM reference module flows.

Reviewers: vkuzo, supriyar, jcaip

Subscribers: vkuzo, supriyar, jcaip

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Create new FX-based LSTM reference module"


Summary: The previous LSTM reference module implementation did
not handle dtypes other than quint8 correctly. This is because
the internal LSTM custom module quantization used eager mode,
which did not insert the q-dq ops properly. E.g., we want the
following reference quantized model:

```
[dq -> linear1_fp32 -> q_to_qint32] -> dq -> q_to_quint8 ->
  [dq - linear2_fp32 -> q_to_quint8] -> dq -> ...
```

This requires two sets of `q - dq` pairs between two adjacent
ops that have different dtypes (linear1 and linear2). However,
these `q - dq` pairs were not inserted in the old flow, because
eager mode required users to insert Quant/DeQuantStubs manually.

This commit changes the internal LSTM custom module quantization
to use FX graph mode quantization, which automatically inserts
the `q - dq` ops that convert the dtypes between adjacent ops
correctly. However, using FX graph mode quantization here comes
with its own set of challenges that required some hacks to get
the end-to-end flow to work. These hacks are detailed in the
comments in the util functions.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_static_lstm_with_custom_fixed_qparams

This commit also updates the corresponding test to verify the
dtypes as well as the qparams in the reference quantized graph.
This test case should serve as an example for users to set up
their own LSTM reference module flows.

Reviewers: vkuzo, supriyar, jcaip

Subscribers: vkuzo, supriyar, jcaip

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Create new FX-based LSTM reference module"


Summary: The previous LSTM reference module implementation did
not handle dtypes other than quint8 correctly. This is because
the internal LSTM custom module quantization used eager mode,
which did not insert the q-dq ops properly. E.g., we want the
following reference quantized model:

```
[dq -> linear1_fp32 -> q_to_qint32] -> dq -> q_to_quint8 ->
  [dq - linear2_fp32 -> q_to_quint8] -> dq -> ...
```

This requires two sets of `q - dq` pairs between two adjacent
ops that have different dtypes (linear1 and linear2). However,
these `q - dq` pairs were not inserted in the old flow, because
eager mode required users to insert Quant/DeQuantStubs manually.

This commit changes the internal LSTM custom module quantization
to use FX graph mode quantization, which automatically inserts
the `q - dq` ops that convert the dtypes between adjacent ops
correctly. However, using FX graph mode quantization here comes
with its own set of challenges that required some hacks to get
the end-to-end flow to work. These hacks are detailed in the
comments in the util functions.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_static_lstm_with_custom_fixed_qparams

This commit also updates the corresponding test to verify the
dtypes as well as the qparams in the reference quantized graph.
This test case should serve as an example for users to set up
their own LSTM reference module flows.

Reviewers: vkuzo, supriyar, jcaip

Subscribers: vkuzo, supriyar, jcaip

[ghstack-poisoned]

vkuzo commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -4545,14 +4546,21 @@ def forward(self, inputs: torch.Tensor, h0: torch.Tensor, c0: torch.Tensor):
-        class UserLSTM(torch.ao.nn.quantizable.LSTM):
+        # Construct a BackendConfig that supports qint32 for certain ops
+        qint32_dtype_config = DTypeConfig(input_dtype=torch.qint32, output_dtype=torch.qint32)
+        my_backend_config = get_qnnpack_backend_config()
```
optional nit: might be cleaner to build from scratch.  `get_qnnpack_backend_config` can change, and it's not clear which parts of it are needed to test this PR

andrewor14 replied:
Actually I tried this and it ended up inflating the test code a lot, so I prefer to leave this as is. Also this is likely how our customers will end up using our APIs, so having this here will serve as a practical example

vkuzo replied:
> Also this is likely how our customers will end up using our APIs, so having this here will serve as a practical example

this is unsafe though, I would not recommend our users to do that, I would usually request changes if I see something like that in a production path.  For example, if their backend doesn't support some other op that qnnpack supports, then they will get silently incorrect results.

andrewor14 committed:
Update on "[Quant][fx] Create new FX-based LSTM reference module"


Summary: The previous LSTM reference module implementation did
not handle dtypes other than quint8 correctly. This is because
the internal LSTM custom module quantization used eager mode,
which did not insert the q-dq ops properly. E.g., we want the
following reference quantized model:

```
[dq -> linear1_fp32 -> q_to_qint32] -> dq -> q_to_quint8 ->
  [dq - linear2_fp32 -> q_to_quint8] -> dq -> ...
```

This requires two sets of `q - dq` pairs between two adjacent
ops that have different dtypes (linear1 and linear2). However,
these `q - dq` pairs were not inserted in the old flow, because
eager mode required users to insert Quant/DeQuantStubs manually.

This commit changes the internal LSTM custom module quantization
to use FX graph mode quantization, which automatically inserts
the `q - dq` ops that convert the dtypes between adjacent ops
correctly. However, using FX graph mode quantization here comes
with its own set of challenges that required some hacks to get
the end-to-end flow to work. These hacks are detailed in the
comments in the util functions.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_static_lstm_with_custom_fixed_qparams

This commit also updates the corresponding test to verify the
dtypes as well as the qparams in the reference quantized graph.
This test case should serve as an example for users to set up
their own LSTM reference module flows.

Reviewers: vkuzo, supriyar, jcaip

Subscribers: vkuzo, supriyar, jcaip

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/lstm_utils.py:
```diff
@@ -0,0 +1,160 @@
+        # HACK: Manually replace the activation_post_process following these ops.
+        # This is needed for FloatFunctional ops because there is currently no way
+        # to configure these ops in FX graph mode quantization today. This is because
+        # the FloatFunctional modules simply disappear from the graph after tracing.
```
FloatFunctional gets swapped to FXFloatFunctional here: https://fburl.com/code/8bxfrqzn

I agree, a good fix would be to remove the use of FloatFunctional in the future

andrewor14 replied:
Yes, though FXFloatFunctional also disappears after tracing because it's just a simple wrapper around the torch ops, so we can't access these by the original name

andrewor14 commented on torch/ao/quantization/fx/lstm_utils.py:
```diff
@@ -0,0 +1,160 @@
+        # to configure these ops in FX graph mode quantization today. This is because
+        # the FloatFunctional modules simply disappear from the graph after tracing.
+        # In the future, we should rewrite quantizable LSTM without FloatFunctionals.
+        node_name_to_activation_post_process_ctr = { 
```
I agree. Unfortunately there was no other way because the FXFloatFunctionals are not preserved in the FX graph, so these just look like naked `torch.add` and `torch.mul` ops. I will add a TODO

supriyar replied:
can we add a check somewhere to ensure that the adds and muls in the LSTM module actually have these name names? Reporting a warning if they are different might be useful if the graph has unexpected names in practice to make debugging easier.

vkuzo replied:
I think just looking at the function target and using a counter will be cleaner.  Since this is LSTM and we know that FX graphs have an order, that is a lot less likely to break compared to node names.

andrewor14 replied:
Ok, I changed this to do the target + counter thing.

supriyar commented on torch/ao/nn/quantizable/modules/rnn.py:
```diff
@@ -83,6 +83,7 @@ def forward(self, x: Tensor, hidden: Optional[Tuple[Tensor, Tensor]] = None) ->
         fgate_cx_igate_cgate = self.fgate_cx_igate_cgate.add(fgate_cx, igate_cgate)
         cy = fgate_cx_igate_cgate
 
+        # TODO: make this tanh a member of the module so its qparams can be configured
```
do we need tanh qparams to be configurable for the current use case?

andrewor14 replied:
Not sure. I tried to keep changes in this file separate for now. If it turns out it's needed I can submit a follow up PR

andrewor14 replied:
Not sure. In this PR I'm trying to isolate the changes to within my util functions. If we need to configure this we can fix this separately in the future.

supriyar commented on torch/ao/quantization/fx/lstm_utils.py:
```diff
@@ -0,0 +1,160 @@
+    for layer in observed_lstm.layers:
+        cell = layer.layer_fw.cell
+        cell = convert_to_reference_fx(cell, backend_config=backend_config)
+        # HACK: Manually remove input quantize nodes and output dequantize nodes,
+        # since custom modules expect quint8 inputs and outputs for now. Note that
```
what does the graph look like if we don't do this? For ex, will we have 2 quantize nodes before the linear op? 
Also, confirming that this means the dtype at input/output boundary to the LSTM module will always be fp32 since we are dealing with reference quantized ops.

andrewor14 replied:
Yes, that was something I spent a long time debugging actually. The problem was because it's a custom module, the caller of this module would feed it quint8 tensors. However, within the module FX graph mode quantization assumes the inputs are fp32, so it'll try to quantize them again, leading to the "Unable to run X with args from QuantizedCPU backend" error. Unfortunately there is no easy way to fix this right now without significant changes to how FX graph mode quantization works.

andrewor14 committed:
Update on "[Quant][fx] Create new FX-based LSTM reference module"


Summary: The previous LSTM reference module implementation did
not handle dtypes other than quint8 correctly. This is because
the internal LSTM custom module quantization used eager mode,
which did not insert the q-dq ops properly. E.g., we want the
following reference quantized model:

```
[dq -> linear1_fp32 -> q_to_qint32] -> dq -> q_to_quint8 ->
  [dq - linear2_fp32 -> q_to_quint8] -> dq -> ...
```

This requires two sets of `q - dq` pairs between two adjacent
ops that have different dtypes (linear1 and linear2). However,
these `q - dq` pairs were not inserted in the old flow, because
eager mode required users to insert Quant/DeQuantStubs manually.

This commit changes the internal LSTM custom module quantization
to use FX graph mode quantization, which automatically inserts
the `q - dq` ops that convert the dtypes between adjacent ops
correctly. However, using FX graph mode quantization here comes
with its own set of challenges that required some hacks to get
the end-to-end flow to work. These hacks are detailed in the
comments in the util functions.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_static_lstm_with_custom_fixed_qparams

This commit also updates the corresponding test to verify the
dtypes as well as the qparams in the reference quantized graph.
This test case should serve as an example for users to set up
their own LSTM reference module flows.

Reviewers: vkuzo, supriyar, jcaip

Subscribers: vkuzo, supriyar, jcaip

[ghstack-poisoned]

andrewor14 committed:
Update on "[Quant][fx] Create new FX-based LSTM reference module"


Summary: The previous LSTM reference module implementation did
not handle dtypes other than quint8 correctly. This is because
the internal LSTM custom module quantization used eager mode,
which did not insert the q-dq ops properly. E.g., we want the
following reference quantized model:

```
[dq -> linear1_fp32 -> q_to_qint32] -> dq -> q_to_quint8 ->
  [dq - linear2_fp32 -> q_to_quint8] -> dq -> ...
```

This requires two sets of `q - dq` pairs between two adjacent
ops that have different dtypes (linear1 and linear2). However,
these `q - dq` pairs were not inserted in the old flow, because
eager mode required users to insert Quant/DeQuantStubs manually.

This commit changes the internal LSTM custom module quantization
to use FX graph mode quantization, which automatically inserts
the `q - dq` ops that convert the dtypes between adjacent ops
correctly. However, using FX graph mode quantization here comes
with its own set of challenges that required some hacks to get
the end-to-end flow to work. These hacks are detailed in the
comments in the util functions.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_static_lstm_with_custom_fixed_qparams

This commit also updates the corresponding test to verify the
dtypes as well as the qparams in the reference quantized graph.
This test case should serve as an example for users to set up
their own LSTM reference module flows.

Reviewers: vkuzo, supriyar, jcaip

Subscribers: vkuzo, supriyar, jcaip

[ghstack-poisoned]

andrewor14 commented:
Ok, I'm merging this. Thanks for everyone's feedback!

andrewor14 commented:
@pytorchbot merge

andrewor14 commented:
@pytorchbot merge -g

andrewor14 commented:
@pytorchbot merge
